src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs

```diff
@@ -138,37 +139,70 @@ public static OrchardCoreBuilder AddDataAccess(this OrchardCoreBuilder builder)
                 services.AddScoped(sp =>
                 {
                     var store = sp.GetService<IStore>();
-
                     if (store == null)
                     {
                         return null;
                     }
 
                     var session = store.CreateSession();
-
                     var scopedServices = sp.GetServices<IScopedIndexProvider>();
-
                     session.RegisterIndexes(scopedServices.ToArray());
 
                     ShellScope.Current
                         .RegisterBeforeDispose(scope =>
                         {
-                            return scope.ServiceProvider
-                                .GetRequiredService<IDocumentStore>()
-                                .CommitAsync();
+                            return session.SaveChangesAsync();
                         })
                         .AddExceptionHandler((scope, e) =>
                         {
-                            return scope.ServiceProvider
-                                .GetRequiredService<IDocumentStore>()
-                                .CancelAsync();
+                            return session.CancelAsync();
                         });
 
                     return session;
                 });
 
-                services.AddScoped<IDocumentStore, DocumentStore>();
-                services.AddSingleton<IFileDocumentStore, FileDocumentStore>();
+                services.AddScoped<IDocumentStore>(sp =>
+                {
+                    var session = sp.GetService<ISession>();
+                    if (session == null)
+                    {
+                        return null;
+                    }
+
+                    var documentStore = new DocumentStore(session);
+
+                    ShellScope.Current
+                        .RegisterBeforeDispose(scope =>
+                        {
+                            return documentStore.CommitAsync();
+                        })
+                        .AddExceptionHandler((scope, e) =>
+                        {
+                            return documentStore.CancelAsync();
+                        });
+
+                    return documentStore;
+                });
+
+                services.AddSingleton<IDocumentFileStore, FileSystemDocumentStore>();
+
+                services.AddScoped<IFileDocumentStore>(sp =>
```

Comment:
This should not be necessary. And we shouldn't have to call `CommitAsync` from this registration. Instead we should just do `AddScoped<IDocumentStore, FileDocumentStore>()` when you want this implementation. (this will replace the default DocumentStore registration).

But the bigger change I'd like to see is to remove the `.RegisterBeforeDispose()` from here. Instead, in the Update/Create methods of the stores, invoke `ShellScope.BeforeDispose` to inform the scope that it needs to actually save the document (invoke CommitAsync()). And this would be done only once by have a bool flag in each IDocumentStore that knows if the callback has already been registered.